mesonbuild/compilers/compilers.py

```diff
@@ -186,6 +187,22 @@ class CompileCheckMode(enum.Enum):
     LINK = 'link'
 
 
+@dataclass
+class CompileCheckResult(T.Generic[_T]):
+
+    """The result of a Compiler check.
+
+    :param result: Whether the check was successful or not
+    :param cached: Whether the result was retrieved from a cache, defaults to False
+    """
+
+    result: _T
+    cached: bool = False
```

Comment:
Does dataclass provide some additional useful functionality that would separate it from this:

```python
class CompileCheckResult:
    def __init__(self, result, cached=False):
        self.result = result
        self.cached = cached
```

IMHO this is clearer, shorter and easier to read than a `dataclass`. It would also make typing simpler, because currently, from what I can tell, every instance of this is `CompileCheckResult[bool]`. Getting rid of the bool part would reduce clutter.

Reply:
That is longer once you add the style expected newline between `class` and `def`, and that has no type checking applied. Now all of the results are `typing.Any`, which means that the type checker does nothing for you, and you can put anything you want there and mypy will happily tell you everything is fine.

Given:
```python
class CompileCheckResult(Generic[_T]):

    def __init__(self, result: _T, cached: bool = False):
        self.result = result
        self.cached = cached
```
vs
```python
@dataclass
class CompileCheckResult(Generic[_T]):

    result: _T
    cached: bool = False
```
I find the dataclass easier to read.

Reply:
I guess we could drop the use in `get_define`, which would drop the need to be generic, but I'd kinda like to push in the other direction and stop using `tuple[_T, bool]` at all, and always use the `CompileCheckResult[_T]` class.

Reply:
> I find the dataclass easier to read.

I don't. It adds a dependency and complexity because you have to know and remember how those magic declarations turn into an object. If it provided additional functionality then it would be justified, but I'm not really sure that it does for something conceptually this simple.

Reply:
Just to be sure: either of these is a _lot_ better than a tuple.

Reply:
I've changed my mind, although the same problem exists for the other types returning `tuple[_T, bool]`, it's not clear that the `__bool__` behavior we have is really what we want in those cases. Given that we have cases like `str | None` and `int`, it's not clear to me what we want to be `falsy` in those situations. should the `str | None` case be `return self.value is not None`? or should it be `return not self.value`?

Therefore, I've dropped the generic parts, and only handled the cases where we return a `bool`, since the whole point of this is to fix the cases where `tuple[bool, bool]` was always truthy. We can worry about the rest later.